tlslite/tlsrecordlayer.py

```diff
@@ -951,89 +1021,173 @@ def _getNextRecord(self):
-                if self.version >= (3,2): #For TLS 1.1, remove explicit IV
-                    b = b[self._readState.encContext.block_size : ]
-
-                #Check padding
-                paddingGood = True
-                paddingLength = b[-1]
-                if (paddingLength+1) > len(b):
-                    paddingGood=False
-                    totalPaddingLength = 0
-                else:
-                    if self.version == (3,0):
-                        totalPaddingLength = paddingLength+1
-                    elif self.version in ((3,1), (3,2), (3,3)):
-                        totalPaddingLength = paddingLength+1
-                        paddingBytes = b[-totalPaddingLength:-1]
-                        for byte in paddingBytes:
-                            if byte != paddingLength:
-                                paddingGood = False
-                                totalPaddingLength = 0
-                    else:
-                        raise AssertionError()
+        # decrypt
+        blockLength = self._readState.encContext.block_size
+        if len(b) % blockLength != 0:
+            for result in self._sendError(AlertDescription.decryption_failed,
+                                          "Encrypted data must be multiple of "\
+                                          "blocksize"):
+                yield result
 
-            #Decrypt if it's a stream cipher
+        b = self._readState.encContext.decrypt(b)
+        if self.version >= (3, 2): # remove explicit IV
+            b = b[self._readState.encContext.block_size:]
+
+        # Check padding
+        paddingGood = True
+        paddingLength = b[-1]
+        if (paddingLength+1) > len(b):
+            paddingGood = False
+            totalPaddingLength = 0
+        else:
+            if self.version == (3, 0):
+                totalPaddingLength = paddingLength+1
             else:
-                paddingGood = True
-                b = self._readState.encContext.decrypt(b)
-                totalPaddingLength = 0
+                totalPaddingLength = paddingLength+1
+                paddingBytes = b[-totalPaddingLength:-1]
+                for byte in paddingBytes:
+                    if byte != paddingLength:
+                        paddingGood = False
+                        totalPaddingLength = 0
+
+        if not paddingGood:
+            for result in self._sendError(AlertDescription.decryption_failed,
+                                          "Encrypted data does not have valid "\
+                                          "padding"):
+                yield result
 
-            #Check MAC
-            macGood = True
-            macLength = self._readState.macContext.digest_size
-            endLength = macLength + totalPaddingLength
-            if endLength > len(b):
-                macGood = False
+        # Remove padding
+        b = b[:-totalPaddingLength]
+
+        yield b
+
+    def _decryptRecordWithMtE(self, recordType, b):
```

Comment:
while application of padding can be moved outside, the decryption code needs to stay separate as the handling of errors is completely different in EtM and MtE case